packages/react-router-dom/__tests__/trailing-slashes-test.tsx

```diff
@@ -337,62 +338,74 @@ describe("trailing slashes", () => {
 
       expect(window.location.href).toBe("https://remix.run/foo/bar");
 
+      function SafeNavigate({ to }: { to: To }) {
+        let navigate = useNavigate();
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+        React.useEffect(() => navigate(to), [to]);
+        return null;
+      }
+
       act(() => {
         ReactDOM.render(
           <BrowserRouter basename="/foo" window={window}>
             <Routes>
               <Route index element={<h1>👋</h1>} />
-              <Route path="bar" element={<Navigate to="" />} />
+              <Route path="bar" element={<SafeNavigate to="" />} />
             </Routes>
           </BrowserRouter>,
           node
         );
       });
 
-      expect(window.location.href).toBe("https://remix.run/foo");
+      expect(window.location.href).toBe("https://remix.run/foo/bar");
     });
 
-    it("should not include trailing slash via useSearchParams", () => {
-      let window = getWindowImpl("/foo");
+    it("should include trailing slash from the Navigate", () => {
+      let window = getWindowImpl("/foo/bar");
       spyOn(window.history, "pushState").and.callThrough();
 
-      expect(window.location.href).toBe("https://remix.run/foo");
+      expect(window.location.href).toBe("https://remix.run/foo/bar");
 
-      function SetSearchParams() {
-        let [, setSearchParams] = useSearchParams();
-        React.useEffect(
-          () => setSearchParams({ key: "value" }),
-          [setSearchParams]
-        );
-        return <h1>👋</h1>;
+      function SafeNavigate({ to }: { to: To }) {
+        let navigate = useNavigate();
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+        React.useEffect(() => navigate(to), [to]);
+        return null;
       }
 
       act(() => {
         ReactDOM.render(
           <BrowserRouter basename="/foo" window={window}>
             <Routes>
-              <Route index element={<SetSearchParams />} />
+              <Route index element={<h1>👋</h1>} />
+              <Route path="bar" element={<SafeNavigate to="/" />} />
             </Routes>
           </BrowserRouter>,
           node
         );
       });
 
-      expect(window.location.href).toBe("https://remix.run/foo?key=value");
+      expect(window.location.href).toBe("https://remix.run/foo/");
```

Comment:
Hm, yeah I think I could get on board with that given `window.location.pathname`'s behavior on root routes.

Currently, the idea was to leave control entirely in the users hands such that if _either_ `basename` or `to` had a trailing slash it would be reflected on linking to the root - but that was viewed maybe a bit too narrowly in the scope of relative navigation.  For example, assuming `Links` are one level deep:

```jsx
<Router basename="/app"> + <Link to=".."> -> <a href="/app">
<Router basename="/app"> + <Link to="../"> -> <a href="/app/">
<Router basename="/app/"> + <Link to=".."> -> <a href="/app/">
<Router basename="/app/"> + <Link to="../"> -> <a href="/app/">
```

But absolute linking to the root does cause an issue since `to="/"` eliminates eliminates the ability to leave off the root route trailing slash.  So instead, when the destination is the root route, `basename` alone will control trailing slash:

```jsx
<Router basename="/app"> + <Link to=".."> -> <a href="/app">
<Router basename="/app"> + <Link to="../"> -> <a href="/app">
<Router basename="/app/"> + <Link to=".."> -> <a href="/app/">
<Router basename="/app/"> + <Link to="../"> -> <a href="/app/">

// And, more to this test example:
<Router basename="/app"> + <Link to="/"> -> <a href="/app">
<Router basename="/app/"> + <Link to="/"> -> <a href="/app/">
```

Sound right?

Reply:
yeah - that's where this specific scenario gets ambiguous.  `<Link to="/">` technically has a trailing slash so the current code preserves it (overriding the fact that the `basename` doesn't).  I can flip the priority there such that `basename` takes precedence on root links